dali/operators/generic/constant.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/generic/constant.h"
+
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+DALI_SCHEMA(Constant)
+  .DocStr(R"code(Produces a batch of constant tensors.
+
+The floating point input data should be placed in `fdata` argument and integer data in `idata`.
+The data is a flat vector of values or a single scalar. The data is then reshaped according
+to the `shape` argument. If the data is scalar, it will be broadcast to fill the entire shape.
+
+The operator only performs meaningful work at first invocation - subsequent calls will return
+a reference to the same memory.
+
+The operator can be automatically instantiated in Python with a call to
+`dali.types.Constant(value, dtype, shape, layout)`. The value can be a scalar, a tuple, a list or
+a numpy array, in which case the `shape` and `dtype` (if not explicitly overridden), will be
+taken from that array.
+
+64-bit integer and double precision arrays are not supported and will be silently downgraded
+to 32-bit.
+)code")
+  .NumInput(0)
+  .NumOutput(1)
+  .AddOptionalArg<int>("shape", "The desired shape of the output. "
+                                 "If not set, the data is assumed to be 1D",
+                  std::vector<int>())
+  .AddOptionalArg<float>("fdata", "Contents of the constant produced (for floating point types).",
+                                 std::vector<float>())
+  .AddOptionalArg<int>("idata", "Contents of the constant produced (for integer types).",
+                                 std::vector<int>())
+  .AddOptionalArg("dtype", "Type of the output data. If not set, the output is float if `fdata` "
+                           "argument is used and int if `idata` is used.", DALI_NO_TYPE)
+  .AddOptionalArg("layout", "Layout info. If set and not empty, the layout must match the "
+                            "dimensionality of the output.", "");
+
+namespace {
+template <typename Dst, typename Src>
+void FillTensorVector(
+  TensorVector<CPUBackend> &dst, const TensorListShape<> &shape, const std::vector<Src> &src) {
+  dst.SetContiguous(false);
+  dst.Resize(shape);
```

Comment:
This `Resize` + `mutable_data<Dst>` later will actually do a batch size of allocations.

If you really want to save that memory, you probably should do a singe Tensor, and share data with it.

Do we really want to be that clever or should we just go with one big allocation and have the actual TensorList underneath here?

It was supposed to get optimized in some MakeContiguous scenarios.

Reply:
If the constant node is supposed to be copied to the GPU, then it should have had been created with device="gpu" in the first place.

Reply:
As for the first point - I traced in in the debugger, it doesn't allocate any memory because at the point of resize the underlying tensors don't have a type yet.